### ISSUE TITLE
Global filter doesn't run when accessor key points to an object

### ISSUE THREAD

**Issue by francislavoie** [TanStack/table#4673]
### Describe the bug

If the accessor key points to an object in the data, then the global filter never gets invoked, even if I define a custom filter function.

This wasted a few hours of my time trying to track down why my global filter wasn't working. I find it very surprising that it would silently ignore my column.

### Your minimal, reproducible example

https://codesandbox.io/p/sandbox/elegant-morning-i68tj7?file=%2Fsrc%2Fmain.tsx&selection=%5B%7B%22endColumn%22%3A17%2C%22endLineNumber%22%3A125%2C%22startColumn%22%3A17%2C%22startLineNumber%22%3A125%7D%5D

### Steps to reproduce

Type things in the global filter, notice the rows are not filtered. Open the console, see that `console.log()` in `fuzzyFilter` is never called.

### Expected behavior

If the accessor points to an object _and_ there's a filter defined (either via `globalFilterFn` or a `filterFn` on the column), then it should still attempt to filter that column.

Or at the very least, there should be some kind of debug log if debug is turned on to explain why a column isn't getting filtered. There's really no observability to discover why the column is ignored.

The docs also make no mention that the column type would make a difference.

### How often does this bug happen?

Every time

### Platform

Any

### react-table version

8.7.6

### TypeScript version

4.9.4

### Terms & Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I understand that if my bug cannot be reliable reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

**Comment by anthonyma94** [comment]
+1, trying to update to v8 and I run into this issue. Seems like a big oversight.

**Comment by anthonyma94** [comment]
@francislavoie I figured out why. It's very briefly mentioned in the [column def guide](https://tanstack.com/table/v8/docs/guide/column-defs) that the accessor retrieves a *primitive* value, and that's the value used for filtering (and other functions). Once I created an accessor function that transforms the object to a string, global filtering works again.

For my use case, the only object being passed in is a dayjs object, so it's easy enough to transform it into a string (I'm doing that to display on-screen anyways). Your use-case might be different.

Either way, I do think there should be better documentation about this potential issue, or some kind of error message when the accessor value is being accessed.

**Comment by THowland1** [comment]
+1 for this issue

I used to be able to do a global search filter on an array column in v7.8.0, but now I cannot (v8.7.9).

The `FilterFnOption` type (for a `filterFn` of a ColumnDef) allows the string values: `arrIncludes`, 
`arrIncludesAll`, `arrIncludesSome`. Which implies that filtering on array-type column values would work, but it does not.
Also, in the `globalFilterFn` callback passed into `useReactTable`, the function never runs with the `columnId` of the array-type column, so I can't even manually write a filter function that can access the array-type value

**Comment by eliasmelgaco** [comment]
+!

**Comment by basememara** [comment]
This was driving me insane for awhile. My column value is an array and tried adding a `filterFn` function to my column def so I can hook in my filter logic for it. This was after trying to simply set it to `arrIncludes` since that implied it works on an array column. Nothing worked. After stumbling on this thread and flattening my array to a string it started searching the field but isn't a perfect search:

```javascript
{
  header: 'Tags',
  accessorFn: (row) => row.tags.join(','),
  cell: ({ row }) => <Tags values={row.original.tags} />
}
```

**Comment by tannerbaum** [comment]
> +1 for this issue
> 
> I used to be able to do a global search filter on an array column in v7.8.0, but now I cannot (v8.7.9).
> 
> The `FilterFnOption` type (for a `filterFn` of a ColumnDef) allows the string values: `arrIncludes`, `arrIncludesAll`, `arrIncludesSome`. Which implies that filtering on array-type column values would work, but it does not. Also, in the `globalFilterFn` callback passed into `useReactTable`, the function never runs with the `columnId` of the array-type column, so I can't even manually write a filter function that can access the array-type value

Yes exactly, the globalFilterFn never runs and no feedback is provided, so it's impossible to figure out this is the cause without finding this issue. No console logging from within, type warnings, debug settings on the hook can help point the way.

**Comment by chiptus** [comment]
maybe the `AccessorFn` should document somehow that its return value should be primitive

**Comment by FlemmingBehrend** [comment]
Any updates on this?

Or does the TanTable not support objects as values?

**Comment by francislavoie** [comment]
@KevinVandy IMO it's fine if objects aren't considered, but I think it's wrong for the global filter to not work at all if any column is an object. I think it should either throw a error/warning, or skip that column and still filter the other non-object columns.

**Comment by Hephaestian** [comment]
Added some commentary to #4919 that is potentially relevant to this.

**Comment by LiMao00** [comment]
+1

**Comment by Aavato-c** [comment]
Thanks for the discussion. The accessorFn did it for me.

I'm working with a project where we have data that the user can modify, including adding columns. We're using a solution to send the column definitions with the data from the server to the client, sometimes having to parse the data using a accessor-prop on a columdef field object. Looks something like this. Hope this helps someone some day.

```python
# Python pydantic model for a column header
  "last_used": {
            "title": "Last used", # Visible on the column
            "accessor": "last_used", # Way to get the value
            "type": "dateFromUnix", # Used when determening how to render the value
            "display": True, # If it's supposed to be in the table in the first place. Sometimes tags are excluded because they're added differently
            "hidden": False, # Should the col be hidden by default
            "priority": 4, # Order of column
            "editable": False # Is the content editable, e.g. uuid's are not editable
        },...]
``` 


```javascript
// So in the columndef it works like this:
Object.values(Fields).forEach((field: fieldObject) => {
    if (field.hidden) {
      return
    }

    // The basecolumns already have for example a checkbox-column
    baseColumns.push({
      id: field.accessor,
      ...field.fieldAccessor,
      header: ({ column }) => {
        return (
          <Button
            variant="ghost"
            onClick={() => column.toggleSorting(column.getIsSorted() === "asc")}>
            {field.title}
            <ArrowUpDown className="ml-2 h-4 w-4" />
          </Button>
        )
      },
      cell: ({ row }) => {
        return parseDataFormat(field, row); # A custom function parsing the data according to it's type
      },
      enableGlobalFilter: true,
      accessorFn: (row) => {
        return row[field.accessor] # THIS FIXED THE PROBLEM
      }
    }
```
